crabs/detection_tracking/datamodule.py

```diff
@@ -0,0 +1,254 @@
+
+    if len(batch) == 0:
+        return ()
+
+    return tuple(zip(*batch))
+
+
+def get_train_transform(config: dict) -> transforms.Compose:
+    """
+    Get the transform function to apply to an input sample during training.
+    This function returns a composed transformation for use with training data.
+    The following transforms are applied in sequence:
+    1. Apply color jittering with brightness and hue adjustment.
+    2. Apply Gaussian blur with specific kernel size and sigma.
+        For example kernel size of (5, 9) and sigma (0.1, 5.0).
+        This value can be changed in the config file
+    3. Convert the image to a PyTorch tensor.
+
+    Parameters
+    ----------
+    config : dict
+        dictionary containing parameters.
+
+    Returns
+    -------
+    transforms.Compose
+        A composed transformation that includes the specified operations.
+
+    Examples
+    --------
+    >>> train_transform = get_train_transform()
+    >>> dataset = MyDataset(transform=train_transform)
+    """
+    # TODO: testing with different transforms
+    custom_transforms = []
+    custom_transforms.append(
+        transforms.ColorJitter(
+            brightness=config["transform_brightness"],
+            hue=config["transform_hue"],
+        )
+    )
+    custom_transforms.append(
+        transforms.GaussianBlur(
+            kernel_size=config["gaussian_blur_params"]["kernel_size"],
+            sigma=config["gaussian_blur_params"]["sigma"],
+        )
+    )
+    custom_transforms.append(transforms.ToTensor())
+
+    return transforms.Compose(custom_transforms)
+
+
+def get_test_transform() -> transforms.Compose:
+    """
+    Get the transform function to apply to an input sample
+    during test (evaluation) and inference.
+
+    Returns
+    -------
+    transforms.Compose
+        A composed transformation that includes the specified operations.
+    """
+    custom_transforms = []
+    custom_transforms.append(transforms.ToTensor())
+
+    return transforms.Compose(custom_transforms)
+
+
+class CustomDataModule(pl.LightningDataModule):
+    """
+    Data module for handling data loading and preprocessing for object detection models.
+
+    Parameters:
+    -----------
+    main_dir : str
+        The main directory path containing the dataset.
+    annotation : str
+        The filename of the COCO annotation JSON file.
+    config : dict
+        dictionary containing parameters.
+    seed_n : int
+        Random seed for data splitting.
+    """
+
+    def __init__(self, main_dir, annotation, config, seed_n):
+        super().__init__()
+        self.main_dir = main_dir
+        self.annotation = annotation
+        self.config = config
+        self.seed_n = seed_n
+
+    def prepare_data(self):
+        """
+        Function to prepare the data (i.e., data downloading)
+        """
+        pass
+
+    def setup(self, stage=None):
+        """
+        Sets up the data loader for the specified stage
```

Comment:
I think "sets up the dataloader" is still a bit ambiguous, and doesn't add much than the function name...
We could maybe say more specifically: "computes train and test IDs for the dataloader", or  "computes the train/test splits".

I was a bit confused about stage None so did a bit of googling....apparently [I'm not the only one](https://stackoverflow.com/a/73261116) 😅 . It seems it's an historic thing:
> The type is annotated optional and the default value is None for historical reasons. The values will always be one of "fit", "validate", "test", "predict".

Maybe we could remove it?